## Preprocessing

In [48]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [49]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=['EIN','NAME'])

In [50]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [51]:
# Look at APPLICATION_TYPE value counts for binning
application_types_binning = application_df['APPLICATION_TYPE'].value_counts()
application_types_binning

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [52]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = list(application_types_binning[application_types_binning<500].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [53]:
# Look at CLASSIFICATION value counts for binning
classification_binning = application_df['CLASSIFICATION'].value_counts()

In [54]:
# You may find it helpful to look at CLASSIFICATION value counts >1
classification_counts = classification_binning[classification_binning>1]
classification_counts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [56]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = list (classification_binning[classification_binning<1000].index)
classifications_to_replace

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [57]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df = pd.get_dummies(application_df,dtype=float)
application_df.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                     0.0   
1       1   108590              1                     0.0   
2       1     5000              0                     0.0   
3       1     6692              1                     0.0   
4       1   142590              1                     0.0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                   1.0                   0.0                  0.0   
1                   0.0                   0.0                  1.0   
2                   0.0                   0.0                  0.0   
3                   0.0                   0.0                  1.0   
4                   0.0                   0.0                  1.0   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                  0.0                  0.0                  0.0  ...   
1                  0.0                  0.0                  0.0  ...   
2                  0.0                  1.0                  0.0  ...   
3                  0.0                  0.0                  0.0  ...   
4                  0.0                  0.0                  0.0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                0.0                     0.0                       0.0   
1                1.0                     0.0                       0.0   
2                0.0                     0.0                       0.0   
3                0.0                     1.0                       0.0   
4                0.0                     0.0                       1.0   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                 0.0               0.0                     0.0   
1                 0.0               0.0                     0.0   
2                 0.0               0.0                     0.0   
3                 0.0               0.0                     0.0   
4                 0.0               0.0                     0.0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0              0.0                0.0                       1.0   
1              0.0                0.0                       1.0   
2              0.0                0.0                       1.0   
3              0.0                0.0                       1.0   
4              0.0                0.0                       1.0   

   SPECIAL_CONSIDERATIONS_Y  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 44 columns]

In [58]:
# Split our preprocessed data into our features and target arrays
y = application_df['IS_SUCCESSFUL'].values
X = application_df.drop('IS_SUCCESSFUL', axis=1).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state = 42)

In [59]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [60]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len( X_train_scaled[0])
hidden_nodes_layer1=80
hidden_nodes_layer2=30
hidden_nodes_layer3=27
nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation='relu'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation='relu'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_15 (Dense)            (None, 80)                3520      
                                                                 
 dense_16 (Dense)            (None, 30)                2430      
                                                                 
 dense_17 (Dense)            (None, 1)                 31        
                                                                 
Total params: 5981 (23.36 KB)
Trainable params: 5981 (23.36 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [61]:
# Compile the model
nn.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics=['accuracy'])

In [73]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,validation_split=0.005, epochs=100)

Epoch 1/100
800/800 [==============================] - 2s 3ms/step - loss: 0.5463 - accuracy: 0.7336 - val_loss: 0.5407 - val_accuracy: 0.7385
Epoch 2/100
800/800 [==============================] - 2s 2ms/step - loss: 0.5461 - accuracy: 0.7331 - val_loss: 0.5392 - val_accuracy: 0.7396
Epoch 3/100
800/800 [==============================] - 2s 2ms/step - loss: 0.5460 - accuracy: 0.7330 - val_loss: 0.5357 - val_accuracy: 0.7391
Epoch 4/100
800/800 [==============================] - 2s 2ms/step - loss: 0.5455 - accuracy: 0.7333 - val_loss: 0.5362 - val_accuracy: 0.7411
Epoch 5/100
800/800 [==============================] - 2s 2ms/step - loss: 0.5448 - accuracy: 0.7333 - val_loss: 0.5385 - val_accuracy: 0.7388
Epoch 6/100
800/800 [==============================] - 3s 3ms/step - loss: 0.5451 - accuracy: 0.7344 - val_loss: 0.5344 - val_accuracy: 0.7422
Epoch 7/100
800/800 [==============================] - 2s 2ms/step - loss: 0.5440 - accuracy: 0.7345 - val_loss: 0.5335 - val_accuracy: 0.7411

In [74]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5676 - accuracy: 0.7304 - 368ms/epoch - 1ms/step
Loss: 0.5676106810569763, Accuracy: 0.7303789854049683


In [78]:
# Export our model to HDF5 file
from google.colab import files

nn.save('AlphabetSoupCharity.h5')

# Report

## Overview
The purpose of this analysis is to evaluate the performance of a deep learning model to predict whether a charity application will be successful. The model was trained on a dataset of over 34,000 charity applications, with features such as the application type, affiliation, classification, use case, organization type, income, and requested funding amount. The target variable is a binary variable indicating whether the application was successful.

## Report

### Data Preprocessing

#### What variable(s) are the target(s) for your model?
* Target variable: IS_SUCCESSFUL

#### What variable(s) are the features for your model?
* Feature variables: APPLICATION_TYPE, AFFILIATION, CLASSIFICATION, USE_CASE, ORGANIZATION, INCOME_AMT, ASK_AMT

#### What variable(s) should be removed from the input data because they are neither targets nor features?
* Variables removed from the input data: EIN and NAME (non-beneficial ID columns)


### Compiling, Training, and Evaluating the Model


#### How many neurons, layers, and activation functions did you select for your neural network model, and why?

##### Neural network model structure:
* 3 hidden layers with 80, 30, and 27 nodes, respectively
* Output layer with 1 node and a sigmoid activation function

##### Activation functions:
* Rectified linear unit (ReLU) activation function for all hidden layers
* Sigmoid activation function for the output layer

#### Were you able to achieve the target model performance?
Yes, at a satisfactory level.
* Target model performance: Accuracy of 73%

#### What steps did you take in your attempts to increase model performance?
Steps taken to increase model performance:
* Tried different numbers of hidden layers and nodes
* Tried different activation functions
* Tried different optimizers
* Tried regularization techniques


## Summary

The overall results of the deep learning model are satisfactory, with an accuracy of 73%. However, there is still room for improvement. Some additional steps that could be taken to improve the model performance include:

* Collecting more data, especially for rare categories of applications.
* Using a more sophisticated neural network architecture, such as a convolutional neural network (CNN).
* Using transfer learning from a pre-trained model.
Recommendation for a different model

A different model that could be used to solve this classification problem is a support vector machine (SVM). SVMs are a type of machine learning algorithm that can be used for both classification and regression tasks. SVMs work by finding a hyperplane in the feature space that separates the data points into two classes.

SVMs have been shown to be effective for a variety of classification tasks, including predicting the success of charity applications. One advantage of SVMs is that they are relatively insensitive to the dimensionality of the data. This means that SVMs can be used with a large number of features without overfitting the training data.